File: backend/storage.py
import json
import os
import aiofiles
from typing import List, Optional
from datetime import datetime
from models import TrainingDataItem, SentReplyItem, FacebookConfig, WebhookLog

DATA_DIR = "data"
TRAINING_DATA_FILE = os.path.join(DATA_DIR, "training-data.json")
SENT_REPLIES_FILE = os.path.join(DATA_DIR, "sent-replies.json")
FACEBOOK_CONFIG_FILE = os.path.join(DATA_DIR, "facebook-config.json")
WEBHOOK_LOGS_FILE = os.path.join(DATA_DIR, "webhook-logs.json")


def ensure_data_directory():
    """Ensure data directory exists"""
    if not os.path.exists(DATA_DIR):
        os.makedirs(DATA_DIR)


async def get_training_data() -> List[TrainingDataItem]:
    """Get all training data"""
    ensure_data_directory()
    
    if not os.path.exists(TRAINING_DATA_FILE):
        return []
    
    async with aiofiles.open(TRAINING_DATA_FILE, 'r', encoding='utf-8') as f:
        content = await f.read()
        data = json.loads(content)
        return [TrainingDataItem(**item) for item in data]


async def save_training_data(data: List[TrainingDataItem]):
    """Save training data"""
    ensure_data_directory()
    
    json_data = [item.model_dump() for item in data]
    async with aiofiles.open(TRAINING_DATA_FILE, 'w', encoding='utf-8') as f:
        await f.write(json.dumps(json_data, indent=2, ensure_ascii=False))


async def get_sent_replies() -> List[SentReplyItem]:
    """Get all sent replies"""
    ensure_data_directory()
    
    if not os.path.exists(SENT_REPLIES_FILE):
        return []
    
    async with aiofiles.open(SENT_REPLIES_FILE, 'r', encoding='utf-8') as f:
        content = await f.read()
        data = json.loads(content)
        return [SentReplyItem(**item) for item in data]


async def save_sent_replies(data: List[SentReplyItem]):
    """Save sent replies"""
    ensure_data_directory()
    
    json_data = [item.model_dump() for item in data]
    async with aiofiles.open(SENT_REPLIES_FILE, 'w', encoding='utf-8') as f:
        await f.write(json.dumps(json_data, indent=2, ensure_ascii=False))


async def get_facebook_config() -> Optional[FacebookConfig]:
    """Get Facebook configuration"""
    ensure_data_directory()
    
    if not os.path.exists(FACEBOOK_CONFIG_FILE):
        return None
    
    async with aiofiles.open(FACEBOOK_CONFIG_FILE, 'r', encoding='utf-8') as f:
        content = await f.read()
        data = json.loads(content)
        return FacebookConfig(**data)


async def save_facebook_config(config: FacebookConfig):
    """Save Facebook configuration"""
    ensure_data_directory()
    
    async with aiofiles.open(FACEBOOK_CONFIG_FILE, 'w', encoding='utf-8') as f:
        await f.write(json.dumps(config.model_dump(), indent=2, ensure_ascii=False))


async def get_webhook_logs() -> List[WebhookLog]:
    """Get webhook logs"""
    ensure_data_directory()
    
    if not os.path.exists(WEBHOOK_LOGS_FILE):
        return []
    
    async with aiofiles.open(WEBHOOK_LOGS_FILE, 'r', encoding='utf-8') as f:
        content = await f.read()
        data = json.loads(content)
        return [WebhookLog(**item) for item in data]


async def save_webhook_log(log: WebhookLog):
    """Save a webhook log entry"""
    ensure_data_directory()
    
    logs = await get_webhook_logs()
    logs.insert(0, log)  # Add to beginning
    
    # Keep only last 100 logs
    logs = logs[:100]
    
    json_data = [item.model_dump() for item in logs]
    async with aiofiles.open(WEBHOOK_LOGS_FILE, 'w', encoding='utf-8') as f:
        await f.write(json.dumps(json_data, indent=2, ensure_ascii=False))

File: lib/storage.ts
import fs from 'fs/promises'
import path from 'path'

const DATA_DIR = path.join(process.cwd(), 'data')
const DATA_FILE = path.join(DATA_DIR, 'training-data.json')
const SENT_REPLIES_FILE = path.join(DATA_DIR, 'sent-replies.json')
const FACEBOOK_CONFIG_FILE = path.join(DATA_DIR, 'facebook-config.json')
const WEBHOOK_LOGS_FILE = path.join(DATA_DIR, 'webhook-logs.json')

export interface TrainingDataItem {
  id: string
  message: string
  reply: string
  timestamp: string
}

export interface SentReplyItem {
  id: string
  message: string
  reply: string
  mode: 'ai' | 'manual'
  timestamp: string
}

export interface FacebookConfig {
  pageId: string
  pageName: string
  accessToken: string
  verifyToken: string
  webhookUrl: string
  isConnected: boolean
}

export interface WebhookLog {
  timestamp: string
  type: string
  data: any
}

export async function ensureDataDirectory() {
  try {
    await fs.access(DATA_DIR)
  } catch {
    await fs.mkdir(DATA_DIR, { recursive: true })
  }
}

export async function getTrainingData(): Promise<TrainingDataItem[]> {
  await ensureDataDirectory()
  
  try {
    const data = await fs.readFile(DATA_FILE, 'utf-8')
    return JSON.parse(data)
  } catch (error) {
    // If file doesn't exist, return empty array
    return []
  }
}

export async function saveTrainingData(data: TrainingDataItem[]): Promise<void> {
  await ensureDataDirectory()
  await fs.writeFile(DATA_FILE, JSON.stringify(data, null, 2), 'utf-8')
}

export async function getSentReplies(): Promise<SentReplyItem[]> {
  await ensureDataDirectory()
  
  try {
    const data = await fs.readFile(SENT_REPLIES_FILE, 'utf-8')
    return JSON.parse(data)
  } catch (error) {
    // If file doesn't exist, return empty array
    return []
  }
}

export async function saveSentReplies(data: SentReplyItem[]): Promise<void> {
  await ensureDataDirectory()
  await fs.writeFile(SENT_REPLIES_FILE, JSON.stringify(data, null, 2), 'utf-8')
}

export async function getFacebookConfig(): Promise<FacebookConfig | null> {
  await ensureDataDirectory()
  
  try {
    const data = await fs.readFile(FACEBOOK_CONFIG_FILE, 'utf-8')
    return JSON.parse(data)
  } catch (error) {
    return null
  }
}

export async function saveFacebookConfig(config: FacebookConfig): Promise<void> {
  await ensureDataDirectory()
  await fs.writeFile(FACEBOOK_CONFIG_FILE, JSON.stringify(config, null, 2), 'utf-8')
}

export async function getWebhookLogs(): Promise<WebhookLog[]> {
  await ensureDataDirectory()
  
  try {
    const data = await fs.readFile(WEBHOOK_LOGS_FILE, 'utf-8')
    return JSON.parse(data)
  } catch (error) {
    return []
  }
}

export async function saveWebhookLog(log: WebhookLog): Promise<void> {
  await ensureDataDirectory()
  
  const logs = await getWebhookLogs()
  logs.unshift(log) // Add to beginning
  
  // Keep only last 100 logs
  const trimmedLogs = logs.slice(0, 100)
  
  await fs.writeFile(WEBHOOK_LOGS_FILE, JSON.stringify(trimmedLogs, null, 2), 'utf-8')
}
